Plugin/RenderingCacheContextPlugin.php

```diff
@@ -59,4 +58,52 @@ public function afterGetData(HttpContext $subject, $data)
 
         return $data;
     }
+
+    /**
+     * @param int $storeId
+     * @return string
+     */
+    protected function getOriginalRoute(int $storeId): string
+    {
+        $requestUri = $this->request->getRequestUri();
+
+        $rewrite = $this->urlFinder->findOneByData([
+            'request_path' => ltrim($requestUri, '/'),
+            'store_id'     => $storeId,
+        ]);
+
+        return $rewrite?->getTargetPath() ?? "";
+    }
+
+    /**
+     * @param string $path
+     * @return bool
+     */
+    protected function isCategoryRoute(string $path): bool {
+        $path = ltrim($path, '/');
+        return str_starts_with($path, self::CATEGORY_ROUTE);
+    }
+
+    /**
+     * This can be called in a variety of contexts - so this should always be called consistently
+     *
+     * @param int $storeId
+     * @return bool
+     */
+    protected function isCategoryPage(int $storeId): bool
+    {
+        return $this->request->getControllerName() === self::CATEGORY_CONTROLLER
```

Comment:
So if I understood correctly, this is the main difference with the current implementation, getting the controller name can be not sufficient in some cases, that's why we also check the rewrites ?

Reply:
Yes there were 3 scenarios I observed depending on who the request is dispatched 

1. With a controller (on MISS) 
or a URL (on HIT) which can either be 
1. The real path or
1. The rewrite path where we then need to check the original path

The objective in all 3 cases is to correctly discern if the data is being generated on a request for a category page for the entire lifecyle. If we don't derive that correctly then the context changes and the cache is broken.